minifi-api/include/minifi-c/minifi-c.h

```diff
@@ -0,0 +1,228 @@
+
+MINIFI_DECLARE_CONST_HANDLE(MinifiPropertyValidator);
+MINIFI_DECLARE_HANDLE(MinifiFlowFile);
+MINIFI_DECLARE_HANDLE(MinifiLogger);
+MINIFI_DECLARE_HANDLE(MinifiProcessContext);
+MINIFI_DECLARE_HANDLE(MinifiProcessSession);
+MINIFI_DECLARE_HANDLE(MinifiInputStream);
+MINIFI_DECLARE_HANDLE(MinifiOutputStream);
+MINIFI_DECLARE_HANDLE(MinifiConfigure);
+MINIFI_DECLARE_HANDLE(MinifiExtension);
+MINIFI_DECLARE_HANDLE(MinifiPublishedMetrics);
+
+typedef struct MinifiExtensionCreateInfo {
+  MinifiStringView name;
+  MinifiBool(*initialize)(void*, MinifiConfigure);
+  void* user_data;
+} MinifiExtensionCreateInfo;
+
+typedef enum MinifiStatus {
+  MINIFI_SUCCESS = 0,
+  MINIFI_UNKNOWN_ERROR = 1,
+  MINIFI_NOT_SUPPORTED_PROPERTY = 2,
+  MINIFI_DYNAMIC_PROPERTIES_NOT_SUPPORTED = 3,
+  MINIFI_PROPERTY_NOT_SET = 4,
+  MINIFI_VALIDATION_FAILED = 5,
+  MINIFI_PROCESSOR_YIELD = 6
+} MinifiStatus;
+
+typedef struct MinifiProperty {
+  MinifiStringView name;
+  MinifiStringView display_name;
+  MinifiStringView description;
+  MinifiBool is_required;
+  MinifiBool is_sensitive;
+  uint32_t dependent_properties_count;
+  const MinifiStringView* dependent_properties_ptr;
+  uint32_t exclusive_of_properties_count;
+  const MinifiStringView* exclusive_of_property_names_ptr;
+  const MinifiStringView* exclusive_of_property_values_ptr;
+
+  const MinifiStringView* default_value;
+  uint32_t allowed_values_count;
+  const MinifiStringView* allowed_values_ptr;
+  MinifiPropertyValidator validator;
+
+  uint32_t types_count;
+  const MinifiStringView* types_ptr;
+  MinifiBool supports_expression_language;
+} MinifiProperty;
+
+typedef enum MinifiLogLevel {
+  MINIFI_TRACE = 0,
+  MINIFI_DEBUG = 1,
+  MINIFI_INFO = 2,
+  MINIFI_WARNING = 3,
+  MINIFI_ERROR = 4,
+  MINIFI_CRITICAL = 5,
+  MINIFI_OFF = 6
+} MinifiLogLevel;
+
+typedef struct MinifiProcessorMetadata {
+  MinifiStringView uuid;
+  MinifiStringView name;
+  MinifiLogger logger;  // borrowed reference, live until the processor is live
+} MinifiProcessorMetadata;
+
+typedef struct MinifiProcessorCallbacks {
+  OWNED void*(*create)(MinifiProcessorMetadata);
+  void(*destroy)(OWNED void*);
+  MinifiBool(*isWorkAvailable)(void*);
+  void(*restore)(void*, OWNED MinifiFlowFile);
+  MinifiBool(*getTriggerWhenEmpty)(void*);
+  MinifiStatus(*onTrigger)(void*, MinifiProcessContext, MinifiProcessSession);
+  MinifiStatus(*onSchedule)(void*, MinifiProcessContext);
+  void(*onUnSchedule)(void*);
+  OWNED MinifiPublishedMetrics(*calculateMetrics)(void*);
+} MinifiProcessorCallbacks;
+
+typedef struct MinifiProcessorClassDescription {
+  MinifiStringView module_name;
+  MinifiStringView full_name;  // '::'-delimited fully qualified name e.g. 'org::apache::nifi::minifi::GenerateFlowFile'
+  MinifiStringView description;
+  uint32_t class_properties_count;
+  const MinifiProperty* class_properties_ptr;
+  uint32_t dynamic_properties_count;
+  const MinifiDynamicProperty* dynamic_properties_ptr;
```

Comment:
I think we're mixing description and runtime values here. In a processor description/manifest, dynamic properties are only represented as supported/unsupported, while their values are a runtime information. I would remove these members.

Reply:
we need this for docs generation, as they are in a separate section